compiler-utils/src/main/java/com/squareup/anvil/compiler/internal/reference/FunctionReference.kt

```diff
@@ -33,14 +33,40 @@ public sealed class FunctionReference {
 
   public abstract val fqName: FqName
   public abstract val declaringClass: ClassReference
+  public val name: String get() = fqName.shortName().asString()
 
   public val module: AnvilModuleDescriptor get() = declaringClass.module
 
   public abstract val annotations: List<AnnotationReference>
+  public abstract val parameters: List<ParameterReference>
 
   public abstract fun isAbstract(): Boolean
   public abstract fun visibility(): Visibility
 
+  /**
+   * The return type can be null for generic types like `T`. In this case try to resolve the
+   * return type with [resolveGenericReturnOrNull].
+   */
+  public abstract fun returnTypeOrNull(): ClassReference?
```

Comment:
I don't like that this is nullable right now. Same for `ParameterReference`. I'll explore if introducing a new type `TypeReference` could make the situation easier and better.